File: app/src/test/java/com/jameskbride/eventbusmvpdemo/network/NetworkErrorViewPresenterTest.java
package com.jameskbride.eventbusmvpdemo.network;

import com.jameskbride.eventbusmvpdemo.bus.NetworkRequestEvent;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;

import static org.junit.Assert.assertSame;
import static org.mockito.Mockito.verify;
import static org.mockito.MockitoAnnotations.initMocks;

public class NetworkErrorViewPresenterTest {

    @Mock private NetworkErrorViewPresenter.NetworkErrorView view;

    private EventBus eventBus;

    private NetworkErrorViewPresenter subject;
    private NetworkRequestEvent retryRequest;

    @Before
    public void setUp() {
        initMocks(this);
        eventBus = EventBus.getDefault();

        subject = new NetworkErrorViewPresenter(eventBus);
        subject.view = view;

        eventBus.register(this);
    }

    @After
    public void tearDown() {
        eventBus.unregister(this);
    }

    @Test
    public void onRetryItPostsTheNetworkRequestEvent() {
        NetworkRequestEvent networkRequestEvent = new NetworkRequestEvent();

        subject.retry(networkRequestEvent);

        assertSame(networkRequestEvent, this.retryRequest);
    }

    @Test
    public void onRetryItDismissesTheDialog() {
        subject.retry(new NetworkRequestEvent());

        verify(view).dismiss();
    }

    @Subscribe
    public void onNetworkRequestEvent(NetworkRequestEvent networkRequestEvent) {
        this.retryRequest = networkRequestEvent;
    }
}

File: app/src/test/java/com/jameskbride/eventbusmvpdemo/network/NetworkErrorViewPresenterTest.kt
package com.jameskbride.eventbusmvpdemo.network

import com.jameskbride.eventbusmvpdemo.bus.NetworkRequestEvent
import com.nhaarman.mockito_kotlin.verify
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.EventBus
import org.junit.After
import org.junit.Assert.assertSame
import org.mockito.MockitoAnnotations.initMocks
import org.junit.Before
import org.junit.Test
import org.mockito.Mock


class NetworkErrorViewPresenterTest {

    @Mock private lateinit var view: NetworkErrorView

    private lateinit var eventBus: EventBus

    private lateinit var subject: NetworkErrorViewPresenter
    private lateinit var retryRequest: NetworkRequestEvent

    @Before
    fun setUp() {
        initMocks(this)
        eventBus = EventBus.getDefault()

        subject = NetworkErrorViewPresenter(eventBus)
        subject.view = view

        eventBus.register(this)
    }

    @After
    fun tearDown() {
        eventBus.unregister(this)
    }

    @Test
    fun onRetryItPostsTheNetworkRequestEvent() {
        val networkRequestEvent = NetworkRequestEvent()

        subject.retry(networkRequestEvent)

        assertSame(networkRequestEvent, this.retryRequest)
    }

    @Test
    fun onRetryItDismissesTheDialog() {
        subject.retry(NetworkRequestEvent())

        verify(view).dismiss()
    }

    @Subscribe
    fun onNetworkRequestEvent(networkRequestEvent: NetworkRequestEvent) {
        this.retryRequest = networkRequestEvent
    }
}